### PR TITLE
Binary spec tests

### PR DESCRIPTION
**Pull request by kripken**
wip to get the spec tests pass after `wast => binary => wast`.

### CODE REVIEW

**Comment by kripken** [comment]
This isn't done (we'll need a bunch more binary format changes to get the spec tests passing) but I'd like to merge it as I don't want to work too long in a side branch. Anyone want to review?

File: src/wasm-binary.h [hunk 1]
```diff
@@ -390,8 +390,8 @@ class WasmBinaryWriter : public WasmVisitor<WasmBinaryWriter, void> {
   void writeMemory() {
     if (wasm->memory.max == 0) return;
     if (debug) std::cerr << "== writeMemory" << std::endl;
-    o << int8_t(BinaryConsts::Memory) << int8_t(log2(wasm->memory.initial))
-                                      << int8_t(log2(wasm->memory.max))
+    o << int8_t(BinaryConsts::Memory) << int8_t(ceil(log2(wasm->memory.initial)))
+                                      << int8_t(ceil(log2(wasm->memory.max)))
```
Comment (jfbastien) [inline src/wasm-binary.h]: Why add `ceil`? When you use `std::log2()` you get the integral version but it returns a `double`? Aren't the integer values you care about all correctly represented by `double`?

Also, why `int8_t`?

Comment (kripken) [inline src/wasm-binary.h]: For example, `500` bytes of memory would be `8.97` (`log2` gets an integer cast to a float, returns a float). With `ceil` we get `9`. Then in the round trip, we get `512`, a little more. But better than `256` which breaks some spec tests.

`int8_t` is from the v8 binary format.

Comment (jfbastien) [inline src/wasm-binary.h]: I thought the number had to be a power of 2? I guess it makes sense to have `ceil` if it doesn't have to be.

Comment (kripken) [inline src/wasm-binary.h]: The design and spec repos do not require a power of 2, and various tests depend on it. Only the v8 binary format does.

Comment (JSStats) [inline src/wasm-binary.h]: There appears to have been consensus to make these sizes in units of pages. The patch to v8 has 'lgtm' status, and patches have been submitted to sexp-wasm binaryen and the spec. If there is still consensus then it's just a matter of timing.

Comment (kripken) [inline src/wasm-binary.h]: Yes, if that changes then we can update things here as well.

**Comment by kripken** [comment]
Ok, after a lot more fun here, I have most of the tests running, up to `switch.wast` (alphabetically).

This now has a bunch of fixes elsewhere in the code, so I'd like to merge this ASAP before continuing or taking a break.

**Comment by kripken** [comment]
...aaand now they all pass.

**Comment by kripken** [comment]
I doubt anyone wants to review this right now, so merging. If there are fixes we can do them as followups.

**Pull request merged**